### ISSUE TITLE
Building wheel for py-find-1st (PEP 517) ... error

### ISSUE THREAD

**Issue by Maurits14** [freqtrade/freqtrade#3491]
Ubuntu: 20.04
Python: 3.8
Most recent version of Freqtrade

I am trying to install Freqtrade on a new virtualbox. During the installation process the following error occurred: Building wheel for py-find-1st (PEP 517) ... error.

The last time I set up a virtualbox with Freqtrade this did not happen... Could you please help me out with my problem? Thank you very much in advance!

The command I ran:
python3 -m pip install -e .

```
The full output:
Obtaining file:///home/henk/freqtrade
Collecting SQLAlchemy
  Using cached SQLAlchemy-1.3.17-cp38-cp38-manylinux2010_x86_64.whl (1.3 MB)
Collecting TA-Lib
  Using cached TA-Lib-0.4.18.tar.gz (265 kB)
Collecting arrow
  Using cached arrow-0.15.6-py2.py3-none-any.whl (47 kB)
Collecting cachetools
  Using cached cachetools-4.1.0-py3-none-any.whl (10 kB)
Collecting ccxt>=1.24.96
  Using cached ccxt-1.30.6-py2.py3-none-any.whl (1.7 MB)
Collecting colorama
  Using cached colorama-0.4.3-py2.py3-none-any.whl (15 kB)
Collecting jinja2
  Using cached Jinja2-2.11.2-py2.py3-none-any.whl (125 kB)
Collecting jsonschema
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Collecting numpy
  Using cached numpy-1.18.5-cp38-cp38-manylinux1_x86_64.whl (20.6 MB)
Collecting pandas
  Using cached pandas-1.0.4-cp38-cp38-manylinux1_x86_64.whl (10.0 MB)
Collecting prompt-toolkit
  Using cached prompt_toolkit-3.0.5-py3-none-any.whl (351 kB)
Collecting py_find_1st
  Using cached py_find_1st-1.1.4.tar.gz (8.7 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Collecting pycoingecko
  Using cached pycoingecko-1.2.0-py3-none-any.whl (7.0 kB)
Collecting python-rapidjson
  Using cached python_rapidjson-0.9.1-cp38-cp38-manylinux2010_x86_64.whl (1.4 MB)
Collecting python-telegram-bot
  Using cached python_telegram_bot-12.7-py2.py3-none-any.whl (375 kB)
Collecting questionary
  Using cached questionary-1.5.2-py3-none-any.whl (26 kB)
Collecting requests
  Using cached requests-2.23.0-py2.py3-none-any.whl (58 kB)
Processing /home/henk/.cache/pip/wheels/9a/83/ed/da430ea747e6753650e1ecc052d6ae590ce2f91d0b7d05e1bc/sdnotify-0.3.2-py3-none-any.whl
Collecting tabulate
  Using cached tabulate-0.8.7-py3-none-any.whl (24 kB)
Collecting urllib3
  Using cached urllib3-1.25.9-py2.py3-none-any.whl (126 kB)
Processing /home/henk/.cache/pip/wheels/5f/fd/9e/b6cf5890494cb8ef0b5eaff72e5d55a70fb56316007d6dfe73/wrapt-1.12.1-py3-none-any.whl
Collecting python-dateutil
  Using cached python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
Collecting certifi>=2018.1.18
  Using cached certifi-2020.4.5.2-py2.py3-none-any.whl (157 kB)
Collecting cryptography>=2.6.1
  Using cached cryptography-2.9.2-cp35-abi3-manylinux2010_x86_64.whl (2.7 MB)
Collecting aiohttp==3.6.2; python_version >= "3.5.2"
  Using cached aiohttp-3.6.2-py3-none-any.whl (441 kB)
Collecting aiodns==1.1.1; python_version >= "3.5.2"
  Using cached aiodns-1.1.1-py2.py3-none-any.whl (6.4 kB)
Processing /home/henk/.cache/pip/wheels/82/ad/ec/51f1c7833ed931c471836a6f1afc7c2664bd898e0b4ce30d25/yarl-1.1.0-py3-none-any.whl
Requirement already satisfied: setuptools>=38.5.1 in /home/henk/.env/lib/python3.8/site-packages (from ccxt>=1.24.96->freqtrade===develop-761407f7) (44.0.0)
Collecting MarkupSafe>=0.23
  Using cached MarkupSafe-1.1.1-cp38-cp38-manylinux1_x86_64.whl (32 kB)
Collecting six>=1.11.0
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Processing /home/henk/.cache/pip/wheels/17/be/0f/727fb20889ada6aaaaba861f5f0eb21663533915429ad43f28/pyrsistent-0.16.0-cp38-cp38-linux_x86_64.whl
Collecting attrs>=17.4.0
  Using cached attrs-19.3.0-py2.py3-none-any.whl (39 kB)
Collecting pytz>=2017.2
  Using cached pytz-2020.1-py2.py3-none-any.whl (510 kB)
Collecting wcwidth
  Using cached wcwidth-0.2.4-py2.py3-none-any.whl (30 kB)
Processing /home/henk/.cache/pip/wheels/88/79/e5/598ba17e85eccf2626eab62e4ee8452895636cd542650d450d/tornado-6.0.4-cp38-cp38-linux_x86_64.whl
Collecting future>=0.16.0
  Using cached future-0.18.2.tar.gz (829 kB)
Collecting decorator>=4.4.0
  Using cached decorator-4.4.2-py2.py3-none-any.whl (9.2 kB)
Collecting chardet<4,>=3.0.2
  Using cached chardet-3.0.4-py2.py3-none-any.whl (133 kB)
Collecting idna<3,>=2.5
  Using cached idna-2.9-py2.py3-none-any.whl (58 kB)
Collecting cffi!=1.11.3,>=1.8
  Using cached cffi-1.14.0-cp38-cp38-manylinux1_x86_64.whl (409 kB)
Collecting async-timeout<4.0,>=3.0
  Using cached async_timeout-3.0.1-py3-none-any.whl (8.2 kB)
Collecting multidict<5.0,>=4.5
  Using cached multidict-4.7.6-cp38-cp38-manylinux1_x86_64.whl (162 kB)
Collecting pycares>=1.0.0
  Using cached pycares-3.1.1-cp38-cp38-manylinux2010_x86_64.whl (229 kB)
Collecting pycparser
  Using cached pycparser-2.20-py2.py3-none-any.whl (112 kB)
Using legacy setup.py install for TA-Lib, since package 'wheel' is not installed.
Using legacy setup.py install for future, since package 'wheel' is not installed.
Building wheels for collected packages: py-find-1st
  Building wheel for py-find-1st (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/henk/.env/bin/python3 /home/henk/.env/lib/python3.8/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /tmp/tmpzdpaxmq1
       cwd: /tmp/pip-install-n9a4m062/py-find-1st
  Complete output (18 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-x86_64-3.8
  creating build/lib.linux-x86_64-3.8/utils_find_1st
  copying utils_find_1st/__init__.py -> build/lib.linux-x86_64-3.8/utils_find_1st
  running build_ext
  check for clang compiler ... no
  building 'find_1st' extension
  creating build/temp.linux-x86_64-3.8
  creating build/temp.linux-x86_64-3.8/utils_find_1st
  x86_64-linux-gnu-gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -g -fwrapv -O2 -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DNPY_NO_DEPRECATED_API=NPY_1_13_API_VERSION -I/tmp/pip-build-env-4z30mv1n/overlay/lib/python3.8/site-packages/numpy/core/include -I/home/henk/.env/include -I/usr/include/python3.8 -c utils_find_1st/find_1st.cpp -o build/temp.linux-x86_64-3.8/utils_find_1st/find_1st.o
  utils_find_1st/find_1st.cpp:3:10: fatal error: Python.h: Bestand of map bestaat niet
      3 | #include "Python.h"
        |          ^~~~~~~~~~
  compilation terminated.
  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for py-find-1st
Failed to build py-find-1st
ERROR: Could not build wheels for py-find-1st which use PEP 517 and cannot be installed directly
```

**Comment by xmatthias** [comment]
You will most likely need to install python-dev (`sudo apt-get install python3-dev`) - since the error suggests this.

Now to be honest, i've not seen any py_find_1st errors in the last year or so - so this is rather surprising.

**Comment by Maurits14** [comment]
Thank you very much, it worked out.

**Comment by hroff-1902** [comment]
> Now to be honest, i've not seen any py_find_1st errors in the last year or so - so this is rather surprising.

I think it's for python 3.8 (py_find_1st does not provide a prebuilt wheel for python 3.8, I assume)

**Comment by cyber-mouse** [comment]
I am having the same issue. 

Distro: CentOS 8-stream
Python version: 3.8

command i have used: python3 -m pip install -e .

I tried to install python3-dev, found out that there is only python3-devel available on CentOS. I have been searching on internet and found that python3-dev and python3-devel are nearly same packages. 

After installing the python3-devel package, installation still fails with same error message as mentioned in the issue above.

Do you have an idea how this could be solved? @xmatthias 

Thank you

**Comment by xmatthias** [comment]
python with Centos is a pain, especially if you want a specific version.

It's very well possible that python3-devel is NOT python3.8-devel, but python3.7-devel (or even python3.6-devel). The package description would tell.

In the end, you'll need to have matching header files for your python distribution. 

We don't support CentOS with freqtrade (although it'll work just fine once you have it installed) - since it's more "enterpricy" - and usability is sometimes lacking so you'll need to be a part-time linux admin to get things to work.

Best use google to find out how to install the right header files for your python version - or switch to ubuntu, which is more user-focused and easier to use.

**Comment by cyber-mouse** [comment]
Thank you for your reply, it think i am going to switch to Ubuntu or Debian distro.

**Comment by cyber-mouse** [comment]
I got the solution.

After installing Ubuntu I ran in to the same problem. Script stopped at py-find-1st installation.

I installed python3-dev and also python3-wheel, this helped but i got a different error message, still failing on py-find-1st installation. Then I installed g++ after reading the error message. Installation was sucessfull this time. `sudo apt install g++`

hope this will help  to someone with the same error.

**Comment by xmatthias** [comment]
well technically, we say that you should install build-essential - which should contain g++ ...

**Comment by cgw-9527** [comment]
How did you solve the above problems? I tried these ways you said that the error still exists

**Comment by xmatthias** [comment]
@cgw-9527 please open a new issue if the above solutions don't work for you.

Solutions to all problems above were posted and should work (most often it's the missing `python-dev` package), but since it's multiple different problems, it's unclear what problem you run into.

Also, it's bad practice to highjack an old issue, as it'll notify / ping everyone involved in this issue in the first place.

**Comment by arunvibe** [comment]
ERROR: Could not build wheels for py-find-1st which use PEP 517 and cannot be installed directly 
what  should i do
![Uploading Screenshot (261).png…]()

**Comment by arunvibe** [comment]
ERROR: Could not build wheels for py-find-1st which use PEP 517 and cannot be installed directly
what should i do
![Screenshot (261)](https://user-images.githubusercontent.com/71755657/124142460-15e42480-da3f-11eb-9cb5-8106c141dad3.png)

**Comment by xmatthias** [comment]
You will need to have the build dependencies for py-find-1st - which is most likely the c++ compiler (which can be a pain to install on windows).

if you're having problems with that, i suggest to look at [their repository](https://github.com/roebel/py_find_1st) for further support.

**Comment by cbitsakou** [comment]
> > Now to be honest, i've not seen any py_find_1st errors in the last year or so - so this is rather surprising.
> 
> I think it's for python 3.8 (py_find_1st does not provide a prebuilt wheel for python 3.8, I assume)

Hello,

Do you might know how the issue could be solved for python 3.8 and macOS? I installed the project using the instructions for conda environment, and I get this error. I tried to change the python version to 3.6, 3.7 and 3.9 using this command conda install python=3.x , but I got a message that there is an issue with conflicting packages so the version cannot be changed. Could yo u please help me?

Thank you in advance!

**Comment by xmatthias** [comment]
As mentioned above - best look for support in the [repository of the dependency, in this case py_find_1st](https://github.com/roebel/py_find_1st).
